### DISCUSSION TITLE
Changing channels causes a crash

### DISCUSSION THREAD

**Discussion by DVoorhees1957** [GerryDazoo/Slinger#234] [General]
So I was able to reproduce a problem I noticed a few days ago, which you asked about Gerry. I wasn't doing anything fancy, just changing channels using the "Peanut Remote for TiVO DVR" app on my iPhone. Everything was running local, on the same LAN. Here are the details:

Started slingbox_server.py on my Win7 laptop; everything looked fine

Started a stream in VLC (on my Win7 laptop) from my Slingbox 500; the connected TiVo was on channel 11 at the start

Used the "Peanut Remote for TiVO DVR" app on my iPhone to change channels a few times:
- pressed the channel up key, and the channel changed to 13
- waited about 5 seconds, then pressed the channel down key to go back to channel 11
- waited about 5 seconds, then pressed the channel down key again; the TiVo changed to channel to 9, but screen pixelated horizontally and the software crashed almost immediately

I subsequently went directly to the TiVo in question and verified that, in fact, it had successfully changed to channel 9. I did notice weird little (vertical) artifacts when changing channels directly on the TiVo; could those somehow cause the slingbox server code to freak out?

[Another random thought/question, which might or might not be relevant: Is there any possibility this could be related to the Python 3.10 vs. Python 3.8 issue for Windows 7 issue, which I recently noted in https://github.com/GerryDazoo/Slinger/discussions/213 ?]

Detailed output in the command prompt window:

C:\Tools\Slinger>slingbox_server.py config2.ini
Version : 3.08d Running on Windows-7-6.1.7601-SP1 pid= 3188
Using config file config2.ini
Connection Manager Running on port 8080 with 4 max streams using URL slingbox.
BOXES [('sb350', 'sb350'), ('sb500', 'sb500')]
starting up on  port 8080 8388608
Streamer Running:  4 config2.ini sb350 sb350 8080 8388608
Streamer Running:  4 config2.ini sb500 sb500 8080 8388608
Checking for slingbox at ('192.168.1.104', 5201)
Checking for slingbox at ('192.168.1.102', 5206)
('192.168.1.102', 5206) OK
('192.168.1.104', 5201) OK
sb500 Using slingbox at  ('192.168.1.102', 5206)
sb350 Using slingbox at  ('192.168.1.104', 5201)
Streamer:  sb500 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Streamer:  sb350 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
11/04/2022, 14:20:55.523   Streaming request from ('127.0.0.1', 54993)

Slinginfo  "350/500/M1/M2" 12 60 192.168.1.102 5206 3072 2
sb500 Starting Stream for  127.0.0.1:54993
Connecting... ('192.168.1.102', 5206) Control
sb500 Selecting Video Source 1
Keycodes= [1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 35, 37, 38, 39, 40, 41, 42, 43, 44, 46, 47, 51]
VideoParameters: Resolution= 12 FrameRate= 60 VideoBandwidth= 2000 VideoSmoothness= 63 IframeRate= 5 AudioBitRate= 64
Connecting... ('192.168.1.102', 5206) Stream
sb500 Stream started at 11/04/2022, 14:20:58.072  744 0
.......
11/04/2022, 14:22:14.397  sb500 Stream Terminated for  127.0.0.1:54993 [WinError 10053] An established connection was aborted by the software in your host machine
sb500 Shutting down connections
sb500 Logging Out
Streamer:  sb500 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

My config.ini contents:

[SLINGBOXES]
sb500=sb500
sb350=sb350

[sb500]
sbtype="350/500/M1/M2"
password=xxxxxxxxxxxxxxx
ipaddress=<sb500's local IP address>
port=5206
name=sb500
Resolution=12
FrameRate=60
VideoBandwidth=2000
VideoSmoothness=63
VideoSource=1
maxstreams=2

[sb350]
sbtype="350/500/M1/M2"
password=xxxxxxxxxxxxxxx
ipaddress=<sb350's local IP address>
port=5201
name=sb350
Resolution=12
FrameRate=60
VideoBandwidth=2000
VideoSmoothness=63
VideoSource=1
maxstreams=2

[SERVER]
port=8080
maxremotestreams=4
enableremote=no

**Comment by GerryDazoo** [comment]
I have a theory. but two upfront questions. 
Would you be able to make a patch to the python code to help track this down ?
Could you run a packet capture? Wireshark?
I don't think it is related to the python version.

On your Tivo box do you have an option to fix the video output format? Many STBs have a feature to pass thru to you TV the native video format it is receiving so if a channel is 1080i your TV will get 1080i and if the channel is 720P it will be 720P. I think there's something up when this video format change happens especially when it doesn't match the resolution you've configured in your config.ini.
But the STB also (usually) has the option to fix the video output to 1080I or 720P no matter what the input is. So if you wouldn't mind as a first experiment, have a look at the TIVO config and see if you can fix the video output and make your Resolution in the config.in match and see if the problem persists. This will help narrow down what is happening.

**Reply by DVoorhees1957** [reply]
I can definitely patch Python code; packet capture is a little more foreign to me (I've probably only done it twice in the 20-30 years that I've been building and maintaining PCs)...

Regarding controlling the TiVo video output format: that's certainly possible. There are 5 basic settings under "Video Settings":
* Aspect Correction Mode: Full, Panel or Zoom (was Full, but should have been Panel so I changed it to that)
* Letterbox Color: currently set to Black
* TV Aspect Ratio: 16:9 (selected) or 4:3
* Video Output Format: 480i, 480p, 720p, 1080i, and/or 1080p (pass-thru only); all of these are currently selected
* Video Smoothing: Off (only use On if you see static on some channels, and the picture will lose some sharpness)

As a 1st step, I changed the Aspect Correction Mode to Panel and the Video Output Format to 1080i (only). I assume you'd like me to try to recreate the problem...?

**Reply by DVoorhees1957** [reply]
After making those changes, I repeated the same sequence of steps... changed channels 5 or 6 times, with no crash. I probably won't continue trying to recreate the problem, but I'll add more comments to this discussion if the crash happens again down the road...

**Reply by GerryDazoo** [reply]
No worries about patching the code and doing a packet capture for now.
Can I assume your using VLC? This is starting to smell a lot like the issues we were seeing with the built-in tuner channel changes on the ProHD.

Can you undo you Tivo changes and use SMplayer or ffPlayer for your test?
You may have an audio issue after channel changing but that will confirm what's happening.

**Reply by DVoorhees1957** [reply]
Yes, I'm using VLC. And yes, I can (and will) return the settings to what they were before. I'll first try to recreate the problem using VLC, then I'll try one of those others. (I don't have either one installed on my Win7 machine; do you happen to have links for either one?)

Could changing channels without using Slinger somehow be catching the Slinger code "by surprise"? Which obviously doesn't happen when you change channels with Slinger, since it then knows the channel change is coming...

**Reply by DVoorhees1957** [reply]
Well, recreating the problem with VLC was easy - started the stream, with the TiVo on channel 7... issued a single channel up command via the TiVo remote, and things crashed. So now I need to find SMplayer or ffPlayer...

**Reply by DVoorhees1957** [reply]
OK, I installed SMPLayer on my laptop and opened a stream via that app... changed channels a bunch of times and there was no crash. So does that mean VLC is the culprit?

**Reply by GerryDazoo** [reply]
Yes and no. Try this command line argument for VLC. --no-play-and-pause
It's probably buried down in preferences but I didn't spend much time looking for it.

**Reply by DVoorhees1957** [reply]
Hmm, still crashes. BTW, the "play and pause" option can be set, in fact, via the GUI.; go to Preferences, under "Show Settings" in the lower left, select "All" rather than "Simple", then go to Playlist settings.

**Comment by DVoorhees1957** [comment]
Hi Gerry, is there anything more I need to do to help you diagnose this? The VLC parameter --no-play-and-pause didn't fix things. I could just stop using VLC, but I'm not sure I have a good replacement app on my Win7 laptop... I think you said that SMPlayer can have significant audio lag...

**Reply by GerryDazoo** [reply]
I don't recall saying that but someone else may have. SMPlayer works great in my testing  Give it a try on Win7 and let us know.
